### PR TITLE
feat: Support submissions list filtering

### LINKED ISSUE #250: [FEATURE] Support filtering in Submissions list endpoint
## **Description**
Submissions list endpoint needs to have the ability to filter by `IsComplete`, `Status` and `CompletedAt`. Follow the schema defined in #248 and implement the filtering.

## **Implementation Notes**
Changes are needed in the endpoint request schema. Introduce classes (for the schema changes and it's validation needs) that support the filtering and use them for Submissions list. They can be reused for the filtering of other entities.
Make the necessary changes in the core and DB access logic to return the filtered records.

## **Acceptance Criteria**
- [x] Submissions list endpoint can return data filtered by `IsComplete`, `Status` and `CompletedAt`
- [x] The filtering classes and logic are abstracted where possible so they can be reused for other entities

## **Tasks**
- [x] Submissions list endpoint schema changes
- [x] Apply filtering logic

### PR DESCRIPTION
**Pull request by dspetrov**
# Support submissions list filtering

## Description
Implement general support of filtering in Endatix API following #248. Make the changes for Submissions List endpoint to support filtering.
**Note** It is ready for review, Unit tests are still not added but will be.

## Testing
Using Swagger is on way to test, e.g.:
image
Manual executions with the filters in the query string also will work and any tool for web requests will be helpful.

## Related Issues
closes #250

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the style guidelines of this project

### CODE REVIEW

File: src/Endatix.Core/Specifications/Parameters/FilterCriterion.cs
Comment (PeterMarinov) [reply src/Endatix.Core/Specifications/Parameters/FilterCriterion.cs:85]: Upon checking the rest of the PR I assume your idea is that the validator should do the heavy lifting, so exceptions are less likely. This is good enough for me even though I prefer we don't throw and return the 400 when filter is malformed or not possible. No actions needed

**Review by PeterMarinov — DISMISSED** [review]
I like the PR. I have approved and added directions for discussions. Since this is foundational work with high future impact, I feel it's good to discuss items and log next improvements once we agree on common approach.

File: tests/Endatix.Core.Tests/UseCases/Submissions/ListByFormId/ListByFormIdHandlerTests.cs
Comment (dspetrov) [reply tests/Endatix.Core.Tests/UseCases/Submissions/ListByFormId/ListByFormIdHandlerTests.cs:51]: Yes, thanks. I will.

File: src/Endatix.Core/Specifications/SubmissionsByFormIdSpec.cs
Comment (dspetrov) [reply src/Endatix.Core/Specifications/SubmissionsByFormIdSpec.cs:1]: Do you mean to fix the warnings in SubmissionsByFormIdSpec.cs?
I don's see any warnings. I've seen some proposals from the IDE about the XML comments. I've changed them and there aren't any proposals anymore.

File: src/Endatix.Api/Endpoints/Submissions/ListByFormId.ListByFormIdValidator.cs
Comment (dspetrov) [reply src/Endatix.Api/Endpoints/Submissions/ListByFormId.ListByFormIdValidator.cs:9]: Yes, it's good to have in mind that defining the fields will evolve in the future. Here these are only for validation purposes. It's hard now to say which is the right place to get them defined (e.g. application, tenant, etc) as this will largely depend on the business requirements and needs. I would say we start with the most basic - the type has a set of filterable columns it is the direct owner and we evolve to more flexible approaches depending on the needs.

File: src/Endatix.Core/Specifications/SpecificationExtensions.cs
Comment (dspetrov) [reply src/Endatix.Core/Specifications/SpecificationExtensions.cs:77]: I agree the reflection has its drawback. I'm not sure however, how much it is used around our code base and especially in the libraries we use. I think it's good to have some analysis and let the numbers speak before we put efforts for optimizations. For sure the reflection is slow but who knows it may turn out that compared to a single DB request is almost nothing and optimizing it could improve very little... just an example.
I don't really understand your example from the code you share.
Generally I would suggest we plan and start doing performance testing so we can analyze where are the bottlenecks, then really focus our efforts on what mostly slows our system.
As for caching - this really sounds good and will be an easy way to considerably lower the reflection calls. If you agree, let's start with this step. I can add a task for caching here.

**Comment by dspetrov** [comment]
I've added the tests and it needs a new approval.
While adding the tests I made a few small changes:
- enhanced some of the checks in FilterCriterion
- changed doc comments in a few places
- moved SpecificationExtensions to folder common, which needed some namespace changes
- Extracted the filtering logic from SpecificationExtensions to SpecificationHelper so it can be tested (it's hard to compare in the commit as it's in two files but it's the same code, just moved)
All the above can be seen in the last commit and although includes a number of files, it doesn't change anything in how the implementation works in practice.

@PeterMarinov, please have a look and approve again if any questions don't appear.

File: src/Endatix.Core/Specifications/SpecificationExtensions.cs
Comment (PeterMarinov) [reply src/Endatix.Core/Specifications/SpecificationExtensions.cs]: I have seen apps crashing because of the reflection overhead and hence me being proactive about it. Fully agree on the stress testing instead of blind premature optimizations - this is the way!

For the caching task, yes - please add it. It is an easy win here especially since we don't need cache invalidation (fields don't change unless the code change). For the task, please mark it with  [:zap: performance](https://github.com/endatix/endatix/labels/%3Azap%3A%20performance) and [tech-debt](https://github.com/endatix/endatix/labels/tech-debt).

**Review by PeterMarinov — APPROVED** [review]

File: src/Endatix.Core/Specifications/SpecificationExtensions.cs
Comment (dspetrov) [reply src/Endatix.Core/Specifications/SpecificationExtensions.cs]: Here it is - https://github.com/endatix/endatix/issues/301

**Pull request merged**
